src/mds/MDLog.cc

```diff
@@ -840,6 +840,13 @@ void MDLog::_trim_expired_segments()
 	       << " <= " << ls->seq << "/" << ls->offset << dendl;
       break;
     }
+    //avoid triming if the SP has other references in the MDS   
+    //In this context the SP has 3 references: ls / expired_segments / segments 
+    if ( ls.use_count() > 3 ){ //ugly: mv to const if we want this approch
```

Comment:
In the context of this function the SP has 3 references.
I admit it's ugly even if the comparison will be to const.
another approach can be storing `LogSegment*` as before in this containers:
```
std::set<LogSegment*> expiring_segments;
std::set<LogSegment*> expired_segments; 
```
And the only "legal" count in the MDLog will be from:
```
std::map<uint64_t,LogSegmentRef> segments;
```
@batrick WDYT?

Reply:
> Actually, should it not be 4 since `ls` is also a reference?
> 
Yes, but I think that at this point (`MDLog::_trim_expired_segments` execution) mdlog references are from: `segments` / `expired_segments` / the local `ls`

The reference from `expiring_segments` is being erased before in `MDLog::try_expire`:
https://github.com/ceph/ceph/blob/c88e7ee9e444fd5774ac101a412e733253222a17/src/mds/MDLog.cc#L796-L799  

@batrick am i missing something?

Reply:
no, you're right.